dbt/adapters/redshift/relation_configs/dist.py

```diff
@@ -0,0 +1,164 @@
+)
+from dbt.contracts.graph.nodes import ModelNode
+from dbt.dataclass_schema import StrEnum
+from dbt.exceptions import DbtRuntimeError
+
+from dbt.adapters.redshift.relation_configs.base import RedshiftRelationConfigBase
+
+
+class RedshiftDistStyle(StrEnum):
+    auto = "auto"
+    even = "even"
+    all = "all"
+    key = "key"
+
+    @classmethod
+    def default(cls) -> "RedshiftDistStyle":
+        return cls.auto
+
+
+@dataclass(frozen=True, eq=True, unsafe_hash=True)
+class RedshiftDistConfig(RedshiftRelationConfigBase, RelationConfigValidationMixin):
+    """
+    This config fallows the specs found here:
+    https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_NEW.html
+
+    The following parameters are configurable by dbt:
+    - diststyle: the type of data distribution style to use on the table/materialized view
+    - distkey: the column to use for the dist key if `dist_style` is `key`
+    """
+
+    diststyle: Optional[RedshiftDistStyle] = RedshiftDistStyle.default()
+    distkey: Optional[str] = None
+
+    @property
+    def validation_rules(self) -> Set[RelationConfigValidationRule]:
+        # index rules get run by default with the mixin
+        return {
+            RelationConfigValidationRule(
+                validation_check=not (
+                    self.diststyle == RedshiftDistStyle.key and self.distkey is None
+                ),
+                validation_error=DbtRuntimeError(
+                    "A `RedshiftDistConfig` that specifies a `diststyle` of `key` must provide a value for `distkey`."
+                ),
+            ),
+            RelationConfigValidationRule(
+                validation_check=not (
+                    self.diststyle
+                    in (RedshiftDistStyle.auto, RedshiftDistStyle.even, RedshiftDistStyle.all)
+                    and self.distkey is not None
+                ),
+                validation_error=DbtRuntimeError(
+                    "A `RedshiftDistConfig` that specifies a `distkey` must be of `diststyle` `key`."
+                ),
+            ),
+        }
+
+    @classmethod
+    def from_dict(cls, config_dict) -> "RedshiftDistConfig":
+        kwargs_dict = {
+            "diststyle": config_dict.get("diststyle"),
+            "distkey": config_dict.get("distkey"),
+        }
+        dist: "RedshiftDistConfig" = super().from_dict(kwargs_dict)  # type: ignore
+        return dist
+
+    @classmethod
+    def parse_model_node(cls, model_node: ModelNode) -> dict:
+        """
+        Translate ModelNode objects from the user-provided config into a standard dictionary.
+
+        Args:
+            model_node: the description of the distkey and diststyle from the user in this format:
+
+                {
+                    "dist": any("auto", "even", "all") or "<column_name>"
+                }
+
+        Returns: a standard dictionary describing this `RedshiftDistConfig` instance
+        """
+        dist = model_node.config.extra.get("dist", "")
```

Comment:
```suggestion
        dist = model_node.config.extra.get("dist", "").lower()
```
just a nit but would mean you don't need to call `.lower` a couple times later

Reply:
We merge `diststyle` and `distkey` into one user config of `dist`. In other words, this could be the column which is theoretically possibly case-sensitive. I didn't want to use `.lower()` until I knew it was the `diststyle`.

Less concretely, I also happen to like the separation between "go get the pieces I need" and "do these calculations/formats on the pieces".

Reply:
I compromised by defining it in between. I found myself doing that for the other parser and I liked the consistency. I also removed the error check since we're not, nor can we, do that for the node config version. Because it could be the key, the valid value set is only known at the materialized view level (which eventually could also be table, etc).